<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.361">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Deebul Nair">

<title>devblog - Introduction To Statistical Testing</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">devblog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/" rel="" target=""><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com" rel="" target=""><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Introduction To Statistical Testing</h1>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Deebul Nair </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">Invalid Date</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<section id="introduction-to-statistical-testing" class="level1">
<h1>Introduction To Statistical Testing</h1>
</section>
<section id="introduction-to-statistical-tests" class="level1">
<h1>Introduction To Statistical Tests</h1>
<section id="deebul-nair" class="level2">
<h2 class="anchored" data-anchor-id="deebul-nair">Deebul Nair</h2>
<p>deebul.nair@h-brs.de</p>
</section>
<section id="dr.-teena-hasan" class="level2">
<h2 class="anchored" data-anchor-id="dr.-teena-hasan">Dr.&nbsp;Teena Hasan</h2>
<p>teena.hassan@h-brs.de</p>
<section id="hochschule-bonn-rhein-sieg" class="level3">
<h3 class="anchored" data-anchor-id="hochschule-bonn-rhein-sieg">Hochschule Bonn-Rhein-Sieg</h3>
<section id="human-centered-interaction-in-robotics" class="level4">
<h4 class="anchored" data-anchor-id="human-centered-interaction-in-robotics">Human-Centered Interaction in Robotics</h4>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;skip&quot;}" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.stats <span class="im">as</span> stats</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
</section>
</section>
<section id="learning-goals" class="level1">
<h1>Learning Goals</h1>
<ul>
<li>Research Question vs Research Hypothesis vs Statistical Hypothesis</li>
<li>Why Statistical Test</li>
<li>Intro Null Hypothesis Statistical Testing</li>
<li>Descriptive Statistics</li>
<li>Bunch of Tests</li>
</ul>
</section>
<section id="null-hypothesis-statistical-testing-nhst" class="level1">
<h1>Null Hypothesis Statistical Testing (NHST)?</h1>
<p><img src="images/stages_of_empirial_research.png" class="img-fluid"></p>
<section id="power-analysis" class="level2">
<h2 class="anchored" data-anchor-id="power-analysis">Power Analysis</h2>
<p>References: [1] Guy Hoffman and Xuan Zhao. 2020. A Primer for Conducting Experiments in Human–Robot Interaction. J. Hum.-Robot Interact. 10, 1, Article 6 (March 2021), 31 pages. https://doi.org/10.1145/3412374</p>
</section>
</section>
<section id="this-single-statement-generates-great-robotic-research-1" class="level1">
<h1>This Single Statement Generates Great Robotic Research [1]</h1>
<ul>
<li>[1] https://blog.robotiq.com/this-single-statement-generates-great-robotics-research</li>
</ul>
<blockquote class="blockquote">
<p>A hypothesis, of course. Writing a hypothesis is a key skill for experimental design.</p>
</blockquote>
</section>
<section id="what-is-a-hypothesis-anyway" class="level1">
<h1>What is a Hypothesis Anyway?</h1>
<p>Simply put, a hypothesis is a statement which makes a prediction about something which is not proven. It is a kind of educated guess.</p>
<p>You need a hypothesis when you want to prove something through experimentation.</p>
<ul>
<li>Robotics is a little different from the natural sciences, because it is primarily an engineering field.</li>
<li>The Engineering Process is different from the Scientific Method, because it doesn’t usually make hypotheses and conduct experiments to prove them.</li>
<li>Instead, it defines design criteria and then develops technologies which achieve those criteria.</li>
</ul>
<p>Most of the research work in robotics is spent on developing new technologies and methodologies. However, even though the primary purpose of your research is not to prove a hypothesis, as it is in natural science research, you still need a hypothesis to conduct experiments which prove that your developments work as well as you claim they do.</p>
<p>Therefore, a typical robotics hypothesis might look something like:</p>
<pre><code>"Our algorithm calculates the solution faster than a naive algorithm can calculate the same problem."</code></pre>
</section>
<section id="examples-of-good-hypothesis-from-robotics" class="level1">
<h1>Examples of Good Hypothesis from Robotics</h1>
<section id="multi-robot-grasp-planning-for-sequential-assembly-operations-dogar-et-al-2015" class="level2">
<h2 class="anchored" data-anchor-id="multi-robot-grasp-planning-for-sequential-assembly-operations-dogar-et-al-2015">1. Multi-Robot Grasp Planning for Sequential Assembly Operations (Dogar et al 2015)</h2>
<p>In this paper, the authors use multiple mobile manipulators to construct a flat-packed chair with or without re-grasps (i.e.&nbsp;putting down the piece and picking it up in a different way).</p>
<p>The hypothesis can be more or less expressed like this:</p>
<blockquote class="blockquote">
<p>Our algorithm calculates a good enough manipulation solution with few re-grasps faster than a naive algorithm can calculate the same problem optimally (i.e.&nbsp;with no re-grasps).</p>
</blockquote>
<section id="bad-hypothesis" class="level3">
<h3 class="anchored" data-anchor-id="bad-hypothesis">Bad Hypothesis</h3>
<p>For the sake of demonstration, here is an example of what a less specific hypothesis might look like, one which does not define variables. It is completely unspecific as to what “quickly” means:</p>
<blockquote class="blockquote">
<p>Our algorithm can calculate a solution quickly.</p>
</blockquote>
<p>Another problem might arise if the hypothesis does not include testability. For example, this hypothesis would not be testable, because you could never test all of the other algorithms:</p>
<blockquote class="blockquote">
<p>Our algorithm calculates a solution better than all of the other planning algorithms.</p>
</blockquote>
</section>
</section>
<section id="example-2-grasping-without-squeezing-shear-adhesion-gripper-with-fibrillar-thin-film-hawkes-et-al-2015" class="level2">
<h2 class="anchored" data-anchor-id="example-2-grasping-without-squeezing-shear-adhesion-gripper-with-fibrillar-thin-film-hawkes-et-al-2015">Example 2: Grasping without Squeezing: Shear Adhesion Gripper with Fibrillar Thin Film (Hawkes et al 2015)</h2>
<p>It presents a new type of gripper which, as the authors explain, does not use normal forces to pick up objects as most robotic grippers do. Instead, the gripper uses shear forces which are applied to the manipulated object via a flexible film.</p>
<blockquote class="blockquote">
<p>If a gripper uses shear adhesion forces then it will be a viable option for robotic grasping of objects with a large radius of curvature.</p>
</blockquote>
</section>
</section>
<section id="task-1" class="level1">
<h1>Task 1</h1>
<ul>
<li><p>Write one hypothesis of you RnD/Thesis .</p></li>
<li><p>Discussion</p></li>
</ul>
</section>
<section id="research-hypothesis-vs-statistical-hypothesis" class="level1">
<h1>Research Hypothesis vs Statistical Hypothesis?</h1>
<ul>
<li><p>Statistical hypotheses must be mathematically precise and they must correspond to specific claims about the characteristics of the data generating mechanism (i.e., the “population”).</p></li>
<li><p>Even so, the intent is that statistical hypotheses bear a clear relationship to the substantive research hypotheses that you care about.</p></li>
<li><p>Our hypothesis can be more or less expressed like this: &gt; Our algorithm calculates a good enough manipulation solution with few re-grasps faster than a naive algorithm can calculate the same problem optimally (i.e.&nbsp;with no re-grasps).</p></li>
<li><p>So let’s think about what our statistical hypothesis statement would be.</p></li>
<li><p>The quantity that we are interested in within the experiment are <span class="math inline">\(P(regrasps)_{old}\)</span> and <span class="math inline">\(P(regrasps)_{new}\)</span>, the true-but-unknown probability with which the robot re-grasps in the experiment.</p></li>
</ul>
<p><span class="math display">\[P(regrasps)_{old}  &gt; P(regrasps)_{new}\]</span></p>
</section>
<section id="task-2" class="level1">
<h1>Task 2</h1>
<blockquote class="blockquote">
<p>Convert your hypothesis from above to statistical hypothesis</p>
</blockquote>
</section>
<section id="null-hypothesis-statistical-testing-nhst-1" class="level1">
<h1>Null Hypothesis Statistical Testing (NHST)?</h1>
<p><img src="images/stages_of_empirial_research.png" class="img-fluid"></p>
</section>
<section id="research-questions-and-data-examples" class="level1">
<h1>Research Questions and Data Examples</h1>
<section id="experiment-1-weight-gain" class="level2">
<h2 class="anchored" data-anchor-id="experiment-1-weight-gain">Experiment 1 : Weight Gain</h2>
<section id="references" class="level3">
<h3 class="anchored" data-anchor-id="references">References:</h3>
<ul>
<li>Moore, D. S., McCabe, G. P., and Craig, B. A. (2012). Introduction to the Practice of Statistics (7th ed.). New York: Freeman.</li>
<li>Levine, J. A., Eberhardt, N. L., and Jensen, M. D. (1999) Role of nonexcercise activity thermogenesis in resistance to fat gain in humans. Science, 283:212-214.</li>
</ul>
</section>
<section id="description" class="level3">
<h3 class="anchored" data-anchor-id="description">Description:</h3>
<p>“Weight Gain”, provides weights of 16 participants before and after an eight-week period of excessive calorie intake (Moore et al., 2012, p.&nbsp;425).</p>
</section>
<section id="hypothesis" class="level3">
<h3 class="anchored" data-anchor-id="hypothesis">Hypothesis:</h3>
<p>We will test the hypothesis that 1000 excess calorie intake per day over 8 weeks results in 16 pounds (approximately 7.2 kilograms) weight increase.</p>
</section>
<section id="variables" class="level3">
<h3 class="anchored" data-anchor-id="variables">Variables:</h3>
<pre><code>Weight Before - Weight in pounds (lb) measured before eight weeks of excessive calorie intake.
Weight After - Weight in pounds (lb) measured after eight weeks of excessive calorie intake.
Difference - Weight After - Weight Before.</code></pre>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;subslide&quot;}" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>weight_gain_data <span class="op">=</span> pd.read_csv(<span class="st">'https://raw.githubusercontent.com/jasp-stats/jasp-desktop/4527546659fdbd0849261f22ec84e9db76adc49d/Resources/Data%20Sets/Data</span><span class="sc">%20Li</span><span class="st">brary/2.%20T-Tests/Weight</span><span class="sc">%20G</span><span class="st">ain.csv'</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>weight_gain_data.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="3">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Weight Before</th>
<th data-quarto-table-cell-role="th">Weight After</th>
<th data-quarto-table-cell-role="th">Difference</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>122.54</td>
<td>135.74</td>
<td>13.20</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>120.78</td>
<td>129.36</td>
<td>8.58</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>131.12</td>
<td>145.20</td>
<td>14.08</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>137.06</td>
<td>145.64</td>
<td>8.58</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>163.24</td>
<td>173.80</td>
<td>10.56</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
</section>
</section>
<section id="experiment-2-response-to-eye-color" class="level1">
<h1>Experiment 2 : Response to Eye Color</h1>
<section id="description-1" class="level2">
<h2 class="anchored" data-anchor-id="description-1">Description:</h2>
<p>“Response to Eye Color”, provides post-advertisement attitudes towards a brand expressed by four different groups - each group saw the same advertisement except for the aspect that was manipulated: the eye-color of the model.</p>
</section>
<section id="hypothesis-1" class="level2">
<h2 class="anchored" data-anchor-id="hypothesis-1">Hypothesis:</h2>
<p>The null hypothesis that the attitudes are the same regardless of the eye-color of the model</p>
</section>
<section id="variables-1" class="level2">
<h2 class="anchored" data-anchor-id="variables-1">Variables:</h2>
<pre><code>Group - Experimental conditions (`Blue' = Model with blue eyes, `Brown' = Model with brown eyes, `Green'= Model with green eyes, `Down' = Model's eye color cannot be seen).
Subj - Participant number.
Score - An average of 10 survey questions about attitudes towards the brand (7-point Likert scale). Higher averages correspond to more positive attitudes.</code></pre>
</section>
<section id="data" class="level2">
<h2 class="anchored" data-anchor-id="data">Data</h2>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="4">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>eye_color_data <span class="op">=</span> pd.read_csv(<span class="st">'https://raw.githubusercontent.com/jasp-stats/jasp-desktop/4527546659fdbd0849261f22ec84e9db76adc49d/Resources/Data%20Sets/Data</span><span class="sc">%20Li</span><span class="st">brary/3.%20ANOVA/Response%20to</span><span class="sc">%20E</span><span class="st">ye%20Color.csv'</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>eye_color_data.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Group</th>
<th data-quarto-table-cell-role="th">Subj</th>
<th data-quarto-table-cell-role="th">Score</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>Blue</td>
<td>1</td>
<td>1.3</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>Blue</td>
<td>2</td>
<td>1.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>Blue</td>
<td>3</td>
<td>7.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>Blue</td>
<td>4</td>
<td>4.2</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>Blue</td>
<td>5</td>
<td>5.4</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="references-1" class="level2">
<h2 class="anchored" data-anchor-id="references-1">References:</h2>
<p>Moore, D. S., McCabe, G. P., and Craig, B. A. (2012). Introduction to the Practice of Statistics (7th ed.). New York: Freeman.</p>
<p>Simpson, P. M., Sturges, D. L., and Tanguma, J. (2008). The eyes have it, or do they? The effects of model eye color and eye gaze on consumer as response. The Journal of Applied Business and Economics, 8: 60-72.</p>
</section>
</section>
<section id="experiment-3-brain-size-and-intelligence" class="level1">
<h1>Experiment 3 : Brain Size and Intelligence</h1>
<section id="reference" class="level2">
<h2 class="anchored" data-anchor-id="reference">Reference</h2>
<p>Willerman L, Schultz R, Rutledge JN, Bigler ED. In vivo brain size and intelligence. Intelligence. 1991 Apr 1;15(2):223-8.</p>
</section>
<section id="description-2" class="level2">
<h2 class="anchored" data-anchor-id="description-2">Description</h2>
<p>In this study by Willerman et al.&nbsp;(1991) the researchers use Magnetic Resonance Imaging (MRI) to determine the brain size of the subjects. The researchers take into account gender and body size to draw conclusions about the connection between brain size and intelligence.</p>
</section>
<section id="hypothesis-2" class="level2">
<h2 class="anchored" data-anchor-id="hypothesis-2">Hypothesis</h2>
<p>Are the size and weight of your brain indicators of your mental capacity?</p>
</section>
<section id="variable-names" class="level2">
<h2 class="anchored" data-anchor-id="variable-names">Variable Names:</h2>
<ol type="1">
<li>Gender: Male or Female</li>
<li>FSIQ: Full Scale IQ scores based on the four Wechsler (1981) subtests</li>
<li>VIQ: Verbal IQ scores based on the four Wechsler (1981) subtests</li>
<li>PIQ: Performance IQ scores based on the four Wechsler (1981) subtests</li>
<li>Weight: body weight in pounds</li>
<li>Height: height in inches</li>
<li>MRI_Count: total pixel Count from the 18 MRI scans</li>
</ol>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;subslide&quot;}" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>brain_data <span class="op">=</span> pd.read_csv(<span class="st">'https://scipy-lectures.org/_downloads/brain_size.csv'</span>, sep<span class="op">=</span><span class="st">';'</span>, na_values<span class="op">=</span><span class="st">"."</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>brain_data.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Unnamed: 0</th>
<th data-quarto-table-cell-role="th">Gender</th>
<th data-quarto-table-cell-role="th">FSIQ</th>
<th data-quarto-table-cell-role="th">VIQ</th>
<th data-quarto-table-cell-role="th">PIQ</th>
<th data-quarto-table-cell-role="th">Weight</th>
<th data-quarto-table-cell-role="th">Height</th>
<th data-quarto-table-cell-role="th">MRI_Count</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>1</td>
<td>Female</td>
<td>133</td>
<td>132</td>
<td>124</td>
<td>118.0</td>
<td>64.5</td>
<td>816932</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>2</td>
<td>Male</td>
<td>140</td>
<td>150</td>
<td>124</td>
<td>NaN</td>
<td>72.5</td>
<td>1001121</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>3</td>
<td>Male</td>
<td>139</td>
<td>123</td>
<td>150</td>
<td>143.0</td>
<td>73.3</td>
<td>1038437</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>4</td>
<td>Male</td>
<td>133</td>
<td>129</td>
<td>128</td>
<td>172.0</td>
<td>68.8</td>
<td>965353</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>5</td>
<td>Female</td>
<td>137</td>
<td>132</td>
<td>134</td>
<td>147.0</td>
<td>65.0</td>
<td>951545</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p># Statistical Testing</p>
</section>
</section>
<section id="descriptive-statistics" class="level1">
<h1>1. Descriptive Statistics</h1>
<ul>
<li>counts</li>
<li>mean</li>
<li>median</li>
<li>max</li>
<li>min</li>
<li>….</li>
</ul>
<p><img src="images/descriptive_statistics_metrics.png" class="img-fluid"></p>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;subslide&quot;}" data-execution_count="6">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>brain_data.describe()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Unnamed: 0</th>
<th data-quarto-table-cell-role="th">FSIQ</th>
<th data-quarto-table-cell-role="th">VIQ</th>
<th data-quarto-table-cell-role="th">PIQ</th>
<th data-quarto-table-cell-role="th">Weight</th>
<th data-quarto-table-cell-role="th">Height</th>
<th data-quarto-table-cell-role="th">MRI_Count</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">count</td>
<td>40.000000</td>
<td>40.000000</td>
<td>40.000000</td>
<td>40.00000</td>
<td>38.000000</td>
<td>39.000000</td>
<td>4.000000e+01</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">mean</td>
<td>20.500000</td>
<td>113.450000</td>
<td>112.350000</td>
<td>111.02500</td>
<td>151.052632</td>
<td>68.525641</td>
<td>9.087550e+05</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">std</td>
<td>11.690452</td>
<td>24.082071</td>
<td>23.616107</td>
<td>22.47105</td>
<td>23.478509</td>
<td>3.994649</td>
<td>7.228205e+04</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">min</td>
<td>1.000000</td>
<td>77.000000</td>
<td>71.000000</td>
<td>72.00000</td>
<td>106.000000</td>
<td>62.000000</td>
<td>7.906190e+05</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">25%</td>
<td>10.750000</td>
<td>89.750000</td>
<td>90.000000</td>
<td>88.25000</td>
<td>135.250000</td>
<td>66.000000</td>
<td>8.559185e+05</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">50%</td>
<td>20.500000</td>
<td>116.500000</td>
<td>113.000000</td>
<td>115.00000</td>
<td>146.500000</td>
<td>68.000000</td>
<td>9.053990e+05</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">75%</td>
<td>30.250000</td>
<td>135.500000</td>
<td>129.750000</td>
<td>128.00000</td>
<td>172.000000</td>
<td>70.500000</td>
<td>9.500780e+05</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">max</td>
<td>40.000000</td>
<td>144.000000</td>
<td>150.000000</td>
<td>150.00000</td>
<td>192.000000</td>
<td>77.000000</td>
<td>1.079549e+06</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="mean-vs-median" class="level1">
<h1>Mean vs Median</h1>
<p><img src="images/mean_vs_median.png" class="img-fluid"></p>
<ul>
<li>Learning Statistics with JASP - Page 61</li>
</ul>
</section>
<section id="inter-quantile-region" class="level1">
<h1>Inter Quantile Region</h1>
<ul>
<li>The interquartile range (IQR) is like the range, but instead of the difference between the biggest and smallest value the difference between the 25th percentile and the 75th percentile is taken.</li>
<li>Percentile : Arange the data in asceding order, then 25th percentile is 25% data is less than this value</li>
<li>In fact, we’ve already come across the idea. The median of a data set is its 50th percentile!</li>
</ul>
</section>
<section id="standard-deviation" class="level1">
<h1>Standard Deviation</h1>
<ul>
<li><p>In general, you should expect 68% of the data to fall within 1 standard deviation of the mean,</p></li>
<li><p>95% of the data to fall within 2 standard deviation of the mean, and</p></li>
<li><p>99.7% of the data to fall within 3 standard deviations of the mean.</p></li>
<li><p>This rule tends to work pretty well most of the time, but it’s not exact.</p></li>
<li><p>It’s actually calculated based on an assumption that the histogram is symmetric and “bell shaped”.</p></li>
</ul>
<p><img src="images/sd.png" class="img-fluid"> * Learning Statistics with JASP - Page 64</p>
<section id="plotting" class="level2">
<h2 class="anchored" data-anchor-id="plotting">Plotting</h2>
<ul>
<li>Try plotting the data to get a complete picture</li>
</ul>
<p><img src="images/DinoSequentialSmaller.gif" class="img-fluid"></p>
<p>Matejka, J., &amp; Fitzmaurice, G.W. (2017). Same Stats, Different Graphs: Generating Datasets with Varied Appearance and Identical Statistics through Simulated Annealing. Proceedings of the 2017 CHI Conference on Human Factors in Computing Systems.</p>
</section>
</section>
<section id="plots" class="level1">
<h1>Plots</h1>
<ol type="1">
<li>Scatter plot</li>
<li>Box PLot</li>
<li>Density Plots</li>
<li>Rain cloud plots</li>
</ol>
</section>
<section id="scatter-plot" class="level1">
<h1>Scatter Plot</h1>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="7">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Scater plot </span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co">#todo</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>plt.scatter(<span class="bu">range</span>(weight_gain_data.Difference.size), weight_gain_data.Difference)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<pre><code>&lt;matplotlib.collections.PathCollection at 0x7f0910a22fa0&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="2023-06-15-Intro-To-Statistial-Testing_files/figure-html/cell-7-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;subslide&quot;}" data-execution_count="8">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>plt.scatter( weight_gain_data.Difference, np.zeros(weight_gain_data.Difference.size))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">
<pre><code>&lt;matplotlib.collections.PathCollection at 0x7f090e936910&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="2023-06-15-Intro-To-Statistial-Testing_files/figure-html/cell-8-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="box-plot" class="level1">
<h1>Box Plot</h1>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="9">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Box plot</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>sns.boxplot(x<span class="op">=</span>weight_gain_data[<span class="st">"Difference"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="9">
<pre><code>&lt;AxesSubplot:xlabel='Difference'&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="2023-06-15-Intro-To-Statistial-Testing_files/figure-html/cell-9-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="box-plot-1" class="level1">
<h1>Box plot</h1>
<p><img src="https://chartio.com/assets/79165a/tutorials/charts/box-plots/e30a92b3850d486bef6d0752caaa3d110944b25c9ef3043d821a4998c32bb526/box-plot-construction.png" class="img-fluid"></p>
<ul>
<li>https://chartio.com/learn/charts/box-plot-complete-guide/</li>
</ul>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;subslide&quot;}" data-execution_count="10">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>sns.boxplot(data<span class="op">=</span>weight_gain_data[[<span class="st">"Weight Before"</span>, <span class="st">"Weight After"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="10">
<pre><code>&lt;AxesSubplot:&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="2023-06-15-Intro-To-Statistial-Testing_files/figure-html/cell-10-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;subslide&quot;}" data-execution_count="11">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>sns.violinplot(data<span class="op">=</span>weight_gain_data[[<span class="st">"Weight Before"</span>, <span class="st">"Weight After"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="11">
<pre><code>&lt;AxesSubplot:&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="2023-06-15-Intro-To-Statistial-Testing_files/figure-html/cell-11-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="density-plot" class="level1">
<h1>Density plot</h1>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="12">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>sns.displot(data<span class="op">=</span>weight_gain_data, x<span class="op">=</span><span class="st">"Weight Before"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="2023-06-15-Intro-To-Statistial-Testing_files/figure-html/cell-12-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;slide&quot;}" data-execution_count="13">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>sns.displot(data<span class="op">=</span>weight_gain_data, x<span class="op">=</span><span class="st">"Weight Before"</span>, kde<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="2023-06-15-Intro-To-Statistial-Testing_files/figure-html/cell-13-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;slide&quot;}" data-execution_count="14">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>sns.displot(data<span class="op">=</span>weight_gain_data[[<span class="st">"Weight Before"</span>, <span class="st">"Weight After"</span>]], kind<span class="op">=</span><span class="st">'kde'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="2023-06-15-Intro-To-Statistial-Testing_files/figure-html/cell-14-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="do-you-remember-any-other-plot-seeing-this-plot" class="level1">
<h1>Do you remember any other plot seeing this plot ?</h1>
<p><img src="images/alpha_level.png" class="img-fluid"></p>
</section>
<section id="some-plots-lie" class="level1">
<h1>Some plots lie !</h1>
<p><img src="images/boxplotproblem.png" class="img-fluid"></p>
<p>Matejka, J., &amp; Fitzmaurice, G.W. (2017). Same Stats, Different Graphs: Generating Datasets with Varied Appearance and Identical Statistics through Simulated Annealing. Proceedings of the 2017 CHI Conference on Human Factors in Computing Systems.</p>
</section>
<section id="rain-cloud-plot" class="level1">
<h1>Rain cloud Plot</h1>
<ul>
<li>Combining all the thre plots together.</li>
<li>scatter + density + box plot</li>
</ul>
<p><img src="images/raincloud.png" class="img-fluid"></p>
</section>
<section id="why-statistical-test-why-not-compare-mean" class="level1">
<h1>Why statistical test why not compare mean?</h1>
<ul>
<li>For example our Weight Gain Dataset.</li>
<li>Mean(weight before ) = 144.63</li>
<li>Mean(weight after) = 155.045</li>
<li>Difference of Means = 10.408</li>
</ul>
<p>Can you conclude that there is conclusive evidence that the weights increase with the diet ?</p>
<p>The fundamental question underlying statistical analysis:</p>
<blockquote class="blockquote">
<p>estimating the amount of evidence in support of our hypothesis, even in the presence of noise.</p>
</blockquote>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;subslide&quot;}" data-execution_count="15">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>weight_gain_data.describe()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="15">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Weight Before</th>
<th data-quarto-table-cell-role="th">Weight After</th>
<th data-quarto-table-cell-role="th">Difference</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">count</td>
<td>16.000000</td>
<td>16.000000</td>
<td>16.000000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">mean</td>
<td>144.636250</td>
<td>155.045000</td>
<td>10.408750</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">std</td>
<td>22.704877</td>
<td>21.438057</td>
<td>3.840639</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">min</td>
<td>117.260000</td>
<td>129.360000</td>
<td>3.080000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">25%</td>
<td>126.115000</td>
<td>137.885000</td>
<td>7.700000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">50%</td>
<td>138.270000</td>
<td>147.840000</td>
<td>11.110000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">75%</td>
<td>161.480000</td>
<td>170.335000</td>
<td>13.420000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">max</td>
<td>201.740000</td>
<td>204.820000</td>
<td>15.620000</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="statistical-testing-for-evaluating-experiments" class="level1">
<h1>Statistical Testing for Evaluating Experiments</h1>
<ul>
<li>Deals with the relationship between the value of data, its variance, and the confidence of a conclusion</li>
</ul>
<section id="a-typical-situation" class="level2">
<h2 class="anchored" data-anchor-id="a-typical-situation">A typical situation:</h2>
<ul>
<li>Existing technique A</li>
<li>You developed a new technique B</li>
<li>Key question: Is B better than A?</li>
</ul>
</section>
</section>
<section id="evaluating-experiments" class="level1">
<h1>Evaluating Experiments</h1>
<section id="define-a-performance-measure-e.g." class="level2">
<h2 class="anchored" data-anchor-id="define-a-performance-measure-e.g.">1. Define a performance measure, e.g.</h2>
<ul>
<li>Run-time</li>
<li>Error</li>
<li>Accuracy</li>
<li>Robustness (success rate, MTBF, …)</li>
</ul>
</section>
<section id="collect-data-d" class="level2">
<h2 class="anchored" data-anchor-id="collect-data-d">2. Collect data d</h2>
<ul>
<li>Run both techniques on the data d</li>
<li>How to compare the obtained results <span class="math inline">\(A(d), B(d)\)</span>?</li>
</ul>
</section>
</section>
<section id="st-example" class="level1">
<h1>1st Example</h1>
<section id="scenario" class="level2">
<h2 class="anchored" data-anchor-id="scenario">Scenario</h2>
<ul>
<li>A, B are two HRI techniques</li>
<li>Score is the audience response time</li>
<li>Data d is a given map, start and goal pose</li>
</ul>
</section>
<section id="example" class="level2">
<h2 class="anchored" data-anchor-id="example">Example</h2>
<ul>
<li>A(d) = 0.5 s</li>
<li>B(d) = 0.6 s What does that mean?</li>
</ul>
</section>
</section>
<section id="nd-example" class="level1">
<h1>2nd Example</h1>
<ul>
<li>Same scenario but four tasks</li>
</ul>
<section id="example-1" class="level2">
<h2 class="anchored" data-anchor-id="example-1">Example</h2>
<ul>
<li>A(d) = 0.5 s, 0.4 s, 0.6 s, 0.4 s</li>
<li>B(d) = 0.4 s, 0.3 s, 0.6 s, 0.5 s</li>
</ul>
</section>
<section id="mean-of-the-planning-time-is" class="level2">
<h2 class="anchored" data-anchor-id="mean-of-the-planning-time-is">Mean of the planning time is</h2>
<ul>
<li><span class="math inline">\(μ_A\)</span> = 1.9 s/4 = 0.475 s</li>
<li><span class="math inline">\(μ_B\)</span> = 1.8 s/4 = 0.45 s</li>
</ul>
<p>Is B really better than A?</p>
</section>
</section>
<section id="is-b-better-than-a" class="level1">
<h1>Is B better than A?</h1>
<ul>
<li><span class="math inline">\(μ_A\)</span> = 0.475 s, <span class="math inline">\(μ_B\)</span> = 0.45 s</li>
<li><span class="math inline">\(μ_A\)</span> &gt; <span class="math inline">\(μ_B\)</span>, so B is better than A?!</li>
<li>We just evaluated four tests, thus <span class="math inline">\(μ_A\)</span> and <span class="math inline">\(μ_B\)</span> are rough estimates only</li>
<li>We saw too few data to make statements with high confidence</li>
<li>How can we make a confident statement that B is better than A?</li>
</ul>
</section>
<section id="hypothesis-testing" class="level1">
<h1>Hypothesis Testing</h1>
<blockquote class="blockquote">
<p>“Answer a yes-no question about a population and assess that the answer is wrong.” [Cohen’ 95]</p>
</blockquote>
<ul>
<li>Example: To test that B is different from A, assume they are truly equal.</li>
</ul>
<p>Then, assess the probability of the obtained result. If the probability is small, reject the hypothesis.</p>
</section>
<section id="the-null-hypothesis-h_0" class="level1">
<h1>The Null Hypothesis <span class="math inline">\(H_0\)</span></h1>
<ul>
<li>The null hypothesis is the hypothesis that one wants to reject by analyzing data (from experiments)</li>
<li><span class="math inline">\(H_0\)</span> is the default state</li>
<li>A statistical test can never proof <span class="math inline">\(H_0\)</span></li>
<li>A statistical test can only reject or fail to reject <span class="math inline">\(H_0\)</span></li>
<li>Example: to show that method A is better than B, use <span class="math inline">\(H_0\)</span>: A=B</li>
</ul>
</section>
<section id="typicalpopular-null-hypotheses" class="level1">
<h1>Typical/Popular Null Hypotheses</h1>
<section id="example-1-comparing-one-sample-to-a-known-or-hypothesized-population-mean." class="level3">
<h3 class="anchored" data-anchor-id="example-1-comparing-one-sample-to-a-known-or-hypothesized-population-mean.">Example 1 : Comparing one sample to a known or hypothesized population mean.</h3>
<p><span class="math display">\[ H_0 : \mu = 0 \]</span> $ H_1 : $ (two-tailored test )</p>
<p>$ H_1 : &lt; 0 $ (one-tailored test )</p>
<p>$ H_1 : &gt; 0 $ (one-tailored test )</p>
</section>
<section id="example-2-testing-relationships-between-two-or-more-variables" class="level3">
<h3 class="anchored" data-anchor-id="example-2-testing-relationships-between-two-or-more-variables">Example 2 : Testing relationships between two or more variables</h3>
<p><span class="math display">\[ H_0 : \mu_1 = \mu_2 \]</span> $ H_1 : _1 _2 $ (two-tailored test )</p>
<p>$ H_1 : _1 &lt; _2 $ (one-tailored test )</p>
<p>$ H_1 : _1 &gt; _2 $ (one-tailored test )</p>
</section>
</section>
<section id="p-value" class="level1">
<h1>P Value</h1>
<p><img src="images/pvalue.png" class="img-fluid"></p>
<p>Learning Statistics with JASP, Page 169</p>
</section>
<section id="statistical-tests" class="level1">
<h1>Statistical Tests</h1>
</section>
<section id="students-t-tests" class="level1">
<h1>Student’s t-tests</h1>
<section id="one-sample-location-test" class="level2">
<h2 class="anchored" data-anchor-id="one-sample-location-test">One sample location test</h2>
<p>Research is normally carried out in sample populations, but how close does the sample reflect the whole population? The parametric one-sample t-test determines whether the sample mean is statistically different from a known or hypothesized population mean. The null hypothesis (Ho) tested is that the sample mean is equal to the population mean.</p>
<ul>
<li>Given a µ and σ of a population</li>
<li>Test if a sample (from the population) has a significantly different mean than the population</li>
<li>Sample of size N</li>
</ul>
<p><span class="math display">\[ H_0 : \mu = K \]</span> $ H_1 : K $ (two-tailored test )</p>
<p>$ H_1 : &lt; K $ (one-tailored test )</p>
<p>$ H_1 : &gt; K $ (one-tailored test )</p>
</section>
<section id="sample-t-test-testing-the-value-of-a-population-mean" class="level2">
<h2 class="anchored" data-anchor-id="sample-t-test-testing-the-value-of-a-population-mean">1-sample t-test: testing the value of a population mean</h2>
<p>scipy.stats.ttest_1samp() tests if the population mean of data is likely to be equal to a given value (technically if observations are drawn from a Gaussian distributions of given population mean). It returns the T statistic, and the p-value (see the function’s help):</p>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="16">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy <span class="im">import</span> stats</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>stats.ttest_1samp(brain_data[<span class="st">'VIQ'</span>], <span class="dv">0</span>)   </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="16">
<pre><code>TtestResult(statistic=30.08809997084933, pvalue=1.3289196468727879e-28, df=39)</code></pre>
</div>
</div>
<section id="reportingconclusion" class="level3">
<h3 class="anchored" data-anchor-id="reportingconclusion">Reporting/Conclusion</h3>
<p>With a p-value of 10^-28 we can claim that the population mean for the IQ (VIQ measure) is not 0.</p>
</section>
</section>
</section>
<section id="one-sample-t-test-assumptions" class="level1">
<h1>One Sample t-Test: Assumptions</h1>
<ul>
<li>Independently generated samples</li>
<li>The population distribution is Gaussian (otherwise the t-distribution is not the correct choice)</li>
<li>Mean is known</li>
</ul>
<section id="comments" class="level2">
<h2 class="anchored" data-anchor-id="comments">Comments</h2>
<ul>
<li>The t-Test is quite robust under non-Gaussian distributions</li>
<li>Often a 95% or 99% confidence (=5% or 1% significance) level is used</li>
<li>t-Test is one of the most frequently used tests in science</li>
</ul>
</section>
</section>
<section id="binomial-test" class="level1">
<h1>BINOMIAL TEST</h1>
<ul>
<li>The binomial test is effectively a non-parametric version of the one-sample t-test for use with dichotomous (i.e.&nbsp;yes/no) categorical datasets.</li>
<li>This tests whether or not the sample frequency is statistically different from a known or hypothesized population frequency.</li>
<li>The null hypothesis (Ho) tested is that the sample data frequency is equal to the expected population frequency.</li>
</ul>
<section id="assumptions" class="level2">
<h2 class="anchored" data-anchor-id="assumptions">ASSUMPTIONS</h2>
<p>Three assumptions are required for a binomial test to provide a valid result: * The test variable should be a dichotomous scale (such as yes/no, male/female etc.). * The sample responses should be independent * The sample size is less, but representative of the population</p>
</section>
</section>
<section id="example-laptop" class="level1">
<h1>Example : Laptop</h1>
<ul>
<li><p>Open binomial.csv, this contains one column of data showing the number of students using either a Windows laptop or a MacBook at University.</p></li>
<li><p>In January 2018, when comparing just the two operating systems, the UK market share of Windows was 86% and Mac IOS 14%.</p></li>
<li><p>Null hypothesis <span class="math inline">\(H_0 = 0.86\)</span></p></li>
<li><p>Alternate hypothesis <span class="math inline">\(H_0 \neq 0.86\)</span></p></li>
</ul>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="17">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>laptop_data <span class="op">=</span> pd.read_csv(<span class="st">'Binomial.csv'</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>laptop_data</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>laptop_data.describe()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="17">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Laptop</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">count</td>
<td>90</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">unique</td>
<td>2</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">top</td>
<td>Windows</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">freq</td>
<td>54</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="18">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>laptop_data.value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="18">
<pre><code>Laptop 
Windows    54
Mac        36
dtype: int64</code></pre>
</div>
</div>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;slide&quot;}" data-execution_count="19">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> stats.binomtest(<span class="dv">54</span>, n<span class="op">=</span><span class="dv">90</span>, p<span class="op">=</span><span class="fl">0.86</span>, alternative<span class="op">=</span><span class="st">'two-sided'</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="st">'pvalue : '</span>, result.pvalue)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="st">'95</span><span class="sc">% c</span><span class="st">onfidence level'</span> ,result.proportion_ci(confidence_level<span class="op">=</span><span class="fl">0.95</span>))</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="st">'proportion estimate '</span>, result.proportion_estimate)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>pvalue :  1.1957465473118124e-09
95% confidence level ConfidenceInterval(low=0.4913413183034394, high=0.7018860844296883)
proportion estimate  0.6</code></pre>
</div>
</div>
</section>
<section id="reporting-the-results" class="level1">
<h1>REPORTING THE RESULTS</h1>
<p>The UK proportion of Windows and MacBook users was reported to be 86% and 14% respectively. In a cohort of University students (N=90), a Binomial test revealed that the proportion of students using Windows laptops was significantly less (60%, p&lt;.001) and those using MacBooks significantly more (40.4%, p&lt;.001) than expected.</p>
</section>
<section id="two-sample-t-test" class="level1">
<h1>Two Sample t-Test</h1>
<ul>
<li>Often, one wants to compare the means of two samples to see if both are drawn from populations with equal means</li>
<li>Example: Compare two estimation procedures (operating on potentially different data sets)</li>
<li>Typical Hypotheses <span class="math display">\[ H_0 : \mu_1 = \mu_2 \]</span> $ H_1 : _1 _2 $ (two-tailored test )</li>
</ul>
<p>$ H_1 : _1 &lt; _2 $ (one-tailored test )</p>
<p>$ H_1 : _1 &gt; _2 $ (one-tailored test )</p>
<ul>
<li>This test produces a t-score which is a ration of the differences between the two groups and the differences within the two groups:</li>
<li>A large t-score indicates that there is a greater difference between groups. The smaller the t-score, the more similarity there is between groups.</li>
<li>A t-score of 5 means that the groups are five times as different from each other as they are within each other.</li>
<li>The null hypothesis (Ho) tested is that the population means from the two unrelated groups are equal</li>
</ul>
<section id="sample-t-test-testing-for-difference-across-populations" class="level2">
<h2 class="anchored" data-anchor-id="sample-t-test-testing-for-difference-across-populations">2-sample t-test: testing for difference across populations</h2>
<p>We have seen above that the mean VIQ in the male and female populations were different. To test if this is significant, we do a 2-sample t-test with scipy.stats.ttest_ind():</p>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="20">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>female_viq <span class="op">=</span> brain_data[brain_data[<span class="st">'Gender'</span>] <span class="op">==</span> <span class="st">'Female'</span>][<span class="st">'VIQ'</span>]</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>male_viq <span class="op">=</span> brain_data[brain_data[<span class="st">'Gender'</span>] <span class="op">==</span> <span class="st">'Male'</span>][<span class="st">'VIQ'</span>]</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>stats.ttest_ind(female_viq, male_viq)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="20">
<pre><code>Ttest_indResult(statistic=-0.7726161723275012, pvalue=0.44452876778583217)</code></pre>
</div>
</div>
</section>
<section id="paired-tests-repeated-measurements-on-the-same-individuals" class="level2">
<h2 class="anchored" data-anchor-id="paired-tests-repeated-measurements-on-the-same-individuals">Paired tests: repeated measurements on the same individuals</h2>
<p>PIQ, VIQ, and FSIQ give 3 measures of IQ. Let us test if FISQ and PIQ are significantly different. We can use a 2 sample test:</p>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="21">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>stats.ttest_ind(brain_data[<span class="st">'FSIQ'</span>], brain_data[<span class="st">'PIQ'</span>])   </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="21">
<pre><code>Ttest_indResult(statistic=0.465637596380964, pvalue=0.6427725009414841)</code></pre>
</div>
</div>
<p>The problem with this approach is that it forgets that there are links between observations: FSIQ and PIQ are measured on the same individuals. Thus the variance due to inter-subject variability is confounding, and can be removed, using a “paired test”, or “repeated measures test”:</p>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="22">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>stats.ttest_rel(brain_data[<span class="st">'FSIQ'</span>], brain_data[<span class="st">'PIQ'</span>])   </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="22">
<pre><code>TtestResult(statistic=1.7842019405859857, pvalue=0.08217263818364236, df=39)</code></pre>
</div>
</div>
</section>
</section>
<section id="one-way-anova" class="level1">
<h1>One-way ANOVA</h1>
<ul>
<li>Whereas t-tests compare the means of two groups/conditions, one-way analysis of variance (ANOVA) compares the means of 3 or more groups/conditions.</li>
<li>The null hypothesis tested is that there is no significant difference between the means of all the groups. If the null hypothesis is rejected,.</li>
<li>ANOVA just states that there is a significant difference between the groups but not where those differences occur. * To determine where the group differences are, post hoc (From the Latin post hoc, “after this”) tests are subsequently used.</li>
</ul>
<p><span class="math display">\[H_0 : \mu_1 = \mu_2 = \mu_n \]</span></p>
<section id="assumptions-1" class="level2">
<h2 class="anchored" data-anchor-id="assumptions-1">ASSUMPTIONS</h2>
<p>The independent ANOVA makes the same assumptions as most other parametric tests. * The independent variable must be categorical and the dependent variable must be continuous. * The groups should be independent of each other. * The dependent variable should be approximately normally distributed. * There should be no significant outliers. * There should be homogeneity of variance between the groups otherwise the p-value for the F-statistic may not be reliable.</p>
<ul>
<li>The first 2 assumptions are usually controlled through the use of appropriate research method design.</li>
<li>If the last three assumptions are violated then the non-parametric equivalent, Kruskal-Wallis should be considered instead.</li>
</ul>
<p>Data : ANOVA diets.csv. * This contains A column containing the 3 diets used (A, B and C) * another column containing the absolute amount of weight loss after 8 weeks on one of 3 differentdiets. * For good practice check the descriptive statistics and the boxplots for any extreme outliers</p>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="23">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>diet_loss_data <span class="op">=</span> pd.read_csv(<span class="st">'Independent ANOVA diets.csv'</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>diet_loss_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="23">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Diet</th>
<th data-quarto-table-cell-role="th">Weight loss kg</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>Diet A</td>
<td>3.8</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>Diet A</td>
<td>6.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>Diet A</td>
<td>0.7</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>Diet A</td>
<td>2.9</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>Diet A</td>
<td>2.8</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">67</td>
<td>Diet C</td>
<td>2.8</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">68</td>
<td>Diet C</td>
<td>4.1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">69</td>
<td>Diet C</td>
<td>5.3</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">70</td>
<td>Diet C</td>
<td>9.2</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">71</td>
<td>Diet C</td>
<td>6.1</td>
</tr>
</tbody>
</table>

<p>72 rows × 2 columns</p>
</div>
</div>
</div>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;subslide&quot;}" data-execution_count="24">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>sns.boxplot(data<span class="op">=</span>diet_loss_data, x<span class="op">=</span><span class="st">'Diet'</span>, y<span class="op">=</span><span class="st">'Weight loss kg'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="24">
<pre><code>&lt;AxesSubplot:xlabel='Diet', ylabel='Weight loss kg'&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="2023-06-15-Intro-To-Statistial-Testing_files/figure-html/cell-24-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="25">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>stats.f_oneway(diet_loss_data[diet_loss_data[<span class="st">'Diet'</span>] <span class="op">==</span> <span class="st">'Diet A'</span>][<span class="st">'Weight loss kg'</span>],</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>               diet_loss_data[diet_loss_data[<span class="st">'Diet'</span>] <span class="op">==</span> <span class="st">'Diet B'</span>][<span class="st">'Weight loss kg'</span>],</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>               diet_loss_data[diet_loss_data[<span class="st">'Diet'</span>] <span class="op">==</span> <span class="st">'Diet C'</span>][<span class="st">'Weight loss kg'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">
<pre><code>F_onewayResult(statistic=10.825519115627964, pvalue=8.145779232433143e-05)</code></pre>
</div>
</div>
</section>
</section>
<section id="results" class="level1">
<h1>Results</h1>
<p>The main ANOVA table shows that the F-statistic is significant (p&lt;.001) and that there is a large effect size. Therefore, there is a significant difference between the means of the 3 diet groups.</p>
</section>
<section id="post-hoc-anlysis-tukey_hsd" class="level1">
<h1>Post hoc Anlysis : tukey_hsd</h1>
<ul>
<li><p>Tukey’s honestly significant difference (HSD) test performs pairwise comparison of means for a set of samples.</p></li>
<li><p>Whereas ANOVA (e.g.&nbsp;f_oneway) assesses whether the true means underlying each sample are identical, Tukey’s HSD is a post hoc test used to compare the mean of each sample to the mean of each other sample.</p></li>
<li><p>The null hypothesis is that the distributions underlying the samples all have the same mean.</p></li>
<li><p>The test statistic, which is computed for every possible pairing of samples, is simply the difference between the sample means.</p></li>
<li><p>For each pair, the p-value is the probability under the null hypothesis (and other assumptions; see notes) of observing such an extreme value of the statistic, considering that many pairwise comparisons are being performed.</p></li>
<li><p>Confidence intervals for the difference between each pair of means are also available.</p></li>
<li><p>https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.tukey_hsd.html#scipy.stats.tukey_hsd</p></li>
</ul>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;subslide&quot;}" data-execution_count="29">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>res <span class="op">=</span> stats.tukey_hsd(diet_loss_data[diet_loss_data[<span class="st">'Diet'</span>] <span class="op">==</span> <span class="st">'Diet A'</span>][<span class="st">'Weight loss kg'</span>],</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>               diet_loss_data[diet_loss_data[<span class="st">'Diet'</span>] <span class="op">==</span> <span class="st">'Diet B'</span>][<span class="st">'Weight loss kg'</span>],</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>               diet_loss_data[diet_loss_data[<span class="st">'Diet'</span>] <span class="op">==</span> <span class="st">'Diet C'</span>][<span class="st">'Weight loss kg'</span>])</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Tukey's HSD Pairwise Group Comparisons (95.0% Confidence Interval)
Comparison  Statistic  p-value  Lower CI  Upper CI
 (0 - 1)     -0.404     0.777    -1.832     1.024
 (0 - 2)     -2.579     0.000    -4.007    -1.151
 (1 - 0)      0.404     0.777    -1.024     1.832
 (1 - 2)     -2.175     0.001    -3.603    -0.747
 (2 - 0)      2.579     0.000     1.151     4.007
 (2 - 1)      2.175     0.001     0.747     3.603
</code></pre>
</div>
</div>
</section>
<section id="output" class="level1">
<h1>Output</h1>
<ul>
<li>Post hoc testing shows that there is no significant difference between weight loss on diets A and B.</li>
<li>However, It is significantly higher in diet C compared to diet A (p&lt;.001) and diet B (p=.001). Cohen’s d shows that these differences have a large effect size.</li>
</ul>
</section>
<section id="reporting" class="level1">
<h1>Reporting</h1>
<ul>
<li>Independent one way ANOVA showed a significant effect of the type of diet on weight loss after 10 weeks (F (2, 69) =46.184, p&lt;.001, ω2 = 0.214.</li>
<li>Post hoc testing using Tukey’s correction revealed that diet C resulted in significantly greater weight loss than diet A (p&lt;.001) or diet B (p=.001).</li>
<li>There were no significant differences in weight loss between diets A and B (p=.777)</li>
</ul>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;skip&quot;}" data-execution_count="27">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>pip3 install <span class="op">--</span>upgrade scipy</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Requirement already up-to-date: scipy in /home/deebuls/.local/lib/python3.8/site-packages (1.10.1)
Requirement already satisfied, skipping upgrade: numpy&lt;1.27.0,&gt;=1.19.5 in /home/deebuls/.local/lib/python3.8/site-packages (from scipy) (1.22.4)</code></pre>
</div>
</div>
</section>
<section id="linear-regression" class="level1">
<h1>Linear Regression</h1>
<ul>
<li>This approach is founded on the insight that t-tests and one-way ANOVAs are all special cases of a general linear regression model.</li>
<li>Using regression analysis can help ensure consistency in analyzing and comparing results across studies.</li>
<li>It also offers a straightforward way to control for confounding variables.</li>
<li>Regression tests the null hypothesis (Ho) that there will be no significant prediction of the dependent(outcome) variable by the predictor variable(s).</li>
</ul>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;slide&quot;}" data-execution_count="28">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> sns.scatterplot(data<span class="op">=</span>weight_gain_data, x<span class="op">=</span><span class="st">"Weight Before"</span>, y<span class="op">=</span><span class="st">"Weight After"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="2023-06-15-Intro-To-Statistial-Testing_files/figure-html/cell-28-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="32">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> stats.linregress(weight_gain_data[<span class="st">'Weight Before'</span>], weight_gain_data[<span class="st">'Weight After'</span>])</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(result.intercept, result.intercept_stderr, result.pvalue)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>20.32287010489955 6.04654012186869 2.1185490670472105e-12</code></pre>
</div>
</div>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="34">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>plt.plot(weight_gain_data[<span class="st">'Weight Before'</span>], weight_gain_data[<span class="st">'Weight After'</span>], <span class="st">'o'</span>, label<span class="op">=</span><span class="st">'original data'</span>)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>plt.plot(weight_gain_data[<span class="st">'Weight Before'</span>], result.intercept <span class="op">+</span> result.slope<span class="op">*</span>weight_gain_data[<span class="st">'Weight Before'</span>], <span class="st">'r'</span>, label<span class="op">=</span><span class="st">'fitted line'</span>)</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>plt.legend()</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="2023-06-15-Intro-To-Statistial-Testing_files/figure-html/cell-30-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="assumptions-of-statistical-tests" class="level1">
<h1>Assumptions of Statistical Tests</h1>
<ul>
<li>All of the above-mentioned statistical tests are only valid given certain assumptions.</li>
<li>For example, ANOVA assumes independence of observations, a normal distribution of residuals, and ho- moscedasticity of variances.</li>
<li>There exist statistical methods that check for violations of these assumptions.</li>
</ul>
</section>
<section id="normality-test" class="level1">
<h1>Normality Test</h1>
<ul>
<li><p>Before running the data we have to do the Normality Test</p></li>
<li><p>How to check if the data is normal</p></li>
<li><p>scipy.stats.normal_</p></li>
<li><p>Test whether a sample differs from a normal distribution.</p></li>
<li><p>This function tests the null hypothesis that a sample comes from a normal distribution.</p></li>
<li><p>It is based on D’Agostino and Pearson’s test that combines skew and kurtosis to produce an omnibus test of normality.</p></li>
</ul>
<div class="cell" data-slideshow="{&quot;slide_type&quot;:&quot;fragment&quot;}" data-execution_count="35">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>k2, p <span class="op">=</span> stats.normaltest(diet_loss_data[diet_loss_data[<span class="st">'Diet'</span>] <span class="op">==</span> <span class="st">'Diet A'</span>][<span class="st">'Weight loss kg'</span>])</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> (<span class="st">'pvalue '</span>, p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>pvalue  0.992891828851092</code></pre>
</div>
</div>
</section>
<section id="reporting-1" class="level1">
<h1>Reporting</h1>
<p>p value reported is higher than 0.05 so we have to accept the hypothesis that the data is normally distirbuted</p>
</section>
<section id="other-test" class="level1">
<h1>Other Test</h1>
<ul>
<li>COMPARING MORE THAN TWO RELATED GROUPS - RMANOVA</li>
<li>COMPARING INDEPENDENT GROUPS AND THE EFFECTS OF COVARIATES : ANCOVA
<ul>
<li>ANOVA can be used to compare the means of one variable (dependent) in two or more groups,</li>
<li>whereas analysis of covariance (ANCOVA) sits between ANOVA and regression and compares the means of one (dependent) variable in two or more groups while taking into account the variability of other continuous variables (COVARIATES).</li>
</ul></li>
<li>TWO-WAY INDEPENDENT ANOVA
<ul>
<li>One-way ANOVA tests situations when only one independent variable is manipulated,</li>
<li>two-way ANOVA is used when more than 1 independent variable has been manipulated.</li>
<li>In this case, independent variables are known as factors.</li>
</ul></li>
</ul>
</section>
<section id="bayesian-statistics" class="level1">
<h1>Bayesian Statistics</h1>
<p><img src="bayesian_statistics.png" class="img-fluid"></p>
<p><a href="https://osf.io/zkybj">Bayesian Data Analysis with JASP</a></p>
</section>
<section id="bayesian-statistics-1" class="level1">
<h1>Bayesian Statistics</h1>
<ol type="1">
<li>Bayesian Binomial test</li>
<li>Bayesian ANOVA test</li>
<li>Bayesian Regression test</li>
<li>….</li>
</ol>
</section>
<section id="bayesian-t-test" class="level1">
<h1>Bayesian T-test</h1>
<p><img src="https://best.readthedocs.io/en/latest/_images/smart_drug.png" class="img-fluid"></p>
<p><a href="https://best.readthedocs.io/en/latest/">BEST: Bayesian Estimation Supersedes the t-test</a></p>
</section>
<section id="which-test-should-i-use" class="level1">
<h1>Which Test should I use ?</h1>
<section id="step-1-type-of-test" class="level2">
<h2 class="anchored" data-anchor-id="step-1-type-of-test">Step 1 : Type of test</h2>
<ul>
<li>Comparing one sample to a known or hypothesized population mean.</li>
<li>Testing relationships between two or more variables</li>
<li>Predicting outcomes</li>
<li>Testing for differences between two independent groups</li>
<li>Testing for differences between two related groups</li>
<li>Testing for differences between three or more independent groups</li>
<li>Testing for differences between three or more related groups</li>
<li>Test for interactions between 2 or more independent variables</li>
<li>…</li>
</ul>
</section>
</section>
<section id="comparing-one-sample-to-a-known-or-hypothesized-population-mean." class="level1">
<h1>Comparing one sample to a known or hypothesized population mean.</h1>
<p><img src="onesampletestselection.png" class="img-fluid"></p>
<ul>
<li><a href="https://jasp-stats.org/wp-content/uploads/2020/11/Statistical-Analysis-in-JASP-A-Students-Guide-v14-Nov2020.pdf">Statistical-Analysis-in-JASP-A-Students-Guide-v14-Nov2020</a> Page 165</li>
</ul>
</section>
<section id="task" class="level1">
<h1>Task</h1>
<section id="find-the-statistical-test" class="level2">
<h2 class="anchored" data-anchor-id="find-the-statistical-test">Find the statistical test ?</h2>
<ol type="1">
<li><p>Experiment 1 : Weight Gain ?</p></li>
<li><p>Experiment 2 : Response to eqye color</p></li>
<li><p>Experiment 3: Brain size Inteligence</p></li>
</ol>
</section>
</section>
<section id="what-happens-for-large-n" class="level1">
<h1>What Happens for Large N?</h1>
<ul>
<li><p>Have you seen p-value, statistical test in the papers you have read ?</p></li>
<li><p>Any Deep learning paper ?</p></li>
<li><p>The larger the sample size, the easier it is to show differences…</p></li>
<li><p>… but for large sample sizes, we can show any statistical significant difference no matter how small it is</p></li>
<li><p>A statistically significant difference does not tell anything about if the difference is meaningful!</p></li>
<li><p>See concept of “informativeness”</p></li>
<li><p>For large N, differences may by statistically significant but practically meaningless!</p></li>
</ul>
</section>
<section id="ethical-question" class="level1">
<h1>Ethical Question</h1>
<ul>
<li>Dark side of statistics</li>
<li>https://en.wikipedia.org/wiki/How_to_Lie_with_Statistics</li>
<li>https://genomicsclass.github.io/book/pages/plots_to_avoid.html</li>
<li>https://www.nature.com/articles/s41467-020-19160-7</li>
</ul>
</section>
<section id="conclusion" class="level1">
<h1>Conclusion</h1>
<ul>
<li>When to use NHST?</li>
<li><a href="https://jasp-stats.org/wp-content/uploads/2020/11/Statistical-Analysis-in-JASP-A-Students-Guide-v14-Nov2020.pdf">Statistical-Analysis-in-JASP-A-Students-Guide</a>
<ul>
<li>Select the test you should do</li>
<li>Prepare your experiments and dataset</li>
<li>Do the test</li>
</ul></li>
<li>Dont stop here</li>
<li><a href="http://static.jasp-stats.org/Manuals/Bayesian_Guide_v0_12_2_1.pdf">Statistical Analysis Bayesian Guide Jasp</a></li>
</ul>
</section>
<section id="thank-you" class="level1">
<h1>Thank You</h1>
</section>
<section id="references-2" class="level1">
<h1>References</h1>
<ul>
<li><p>https://scipy-lectures.org/packages/statistics/index.html</p></li>
<li><p>https://ethanweed.github.io/pythonbook/landingpage.html</p></li>
<li><p>https://scientistseessquirrel.wordpress.com/2018/10/02/presenting-statistical-results-pointers-in-scientific-writing/</p></li>
<li><p>https://web.archive.org/web/20091029162244/http://www.wiwi.uni-muenster.de/ioeb/en/organisation/pfaff/stat_overview_table.html</p></li>
<li><p>https://www.jmp.com/en_sg/statistics-knowledge-portal/t-test.html</p></li>
<li><p><a href="https://blog.robotiq.com/this-single-statement-generates-great-robotics-research">Writing Good hypothesis for robotics experiments</a></p></li>
<li><p><a href="http://ais.informatik.uni-freiburg.de/teaching/ws11/robotics2/pdfs/rob2-22-testing.pdf">Hypothesis Testing</a></p></li>
</ul>
</section>
<section id="lab-class" class="level1">
<h1>Lab Class</h1>
<ol type="1">
<li>Statistics with JASP</li>
<li>Exercise 1</li>
</ol>
</section>
<section id="exercise-statiscital-test-1" class="level1">
<h1>Exercise: Statiscital Test 1</h1>
<p>‘Therapeutic Touch’ (TT) is a nursing practice rooted in mysticism but alleged to have a scientific basis. Practitioners of TT claim to treat medical conditions by using their hands to manipulate a ‘human energy field’ perceptible above the patients’ skin. Being a skeptical mind, 9-year old Emily Rosa ventured to test these claims by designing an experiment where TT practitioners had to correctly identify the location of the experimenter’s hand (above their left or right hand) while being blinded by a screen. The results of this experiment were later published in the prestigious Journal of the American Medical Association (Rosa et al., 1998).</p>
<ul>
<li><ol type="1">
<li>Open the “Emily Rosa” dataset https://osf.io/3wg6m/m (under “5. Frequencies”).</li>
</ol></li>
<li><ol start="2" type="1">
<li>Get a descriptive overview of the data by producing a frequency table of the variable Outcome. How many percent of practitioners guessed correctly?</li>
</ol></li>
<li><ol start="3" type="1">
<li>What is the type of test ? Which test will you use ? Explain your answer ?</li>
</ol></li>
<li><ol start="4" type="1">
<li>State the null and alternate huypothesis.</li>
</ol></li>
<li><ol start="5" type="1">
<li>Conduct the test . Report will you accept the Null hypothesis?</li>
</ol></li>
</ul>
</section>
<section id="exercise-statistical-test-2" class="level1">
<h1>Exercise: Statistical Test 2</h1>
<p>A teacher believes that directed reading activities in the classroom can improve the reading ability of elementary school children. She convinces her colleagues to give her the chance to try out the new method on a random sample of 21 third-graders. After they participated for 8 weeks in the program, the children take the Degree of Reading Power test (DRP). Their scores are compared to a control group of 23 children who took the test on the same day and followed the same curriculum apart from the reading activities.</p>
<ul>
<li><ol type="1">
<li>Open the “Directed Reading Activities” or https://osf.io/3wg6m/ (under “2. T-Tests”).</li>
</ol></li>
<li><ol start="2" type="1">
<li>Get a descriptive overview of the data.</li>
</ol>
<ul>
<li><ol type="a">
<li>Create a table that shows the means and standard deviations of DRP scores in the control and treatment group.</li>
</ol></li>
<li><ol start="2" type="a">
<li>Create a boxplot that shows the distribution of DRP scores for each group</li>
</ol></li>
</ul></li>
<li><ol start="3" type="1">
<li>What is the type of test ? Which test will you use ? Explain your answer ?</li>
</ol></li>
<li><ol start="4" type="1">
<li>State the null and alternate huypothesis.</li>
</ol></li>
<li><ol start="5" type="1">
<li>Conduct the test . Report will you accept the Null hypothesis?</li>
</ol></li>
</ul>
</section>
<section id="exercise" class="level1">
<h1>Exercise :</h1>
<p>A researcher is interested in what attributes people look for in a prospective partner. Specifically, the researcher wants to know whether looks or personality play an important role in this selection process, and whether the importance of these factors differs between men and women. They decide to design a speed-dating study where 20 people – 10 men and 10 women – each have speed-dates with 9 other people, who differ in attractiveness (either attractive, average, or ugly) and charisma (either high charisma, some charisma, or no charisma). After each speed date, the participants of the study rate their preference for each date, allowing the researcher to assess the influence of attractiveness and charisma on dating preference, and whether the influence differs between male and female raters. (Note that this is a fictional data set created by Andy Field.)</p>
<ul>
<li><ol type="1">
<li>Open the “Looks or Personality” dataset https://osf.io/3wg6m/ (under “3. ANOVA”).</li>
</ol></li>
<li><ol start="2" type="1">
<li>Create a table that shows the means and standard deviations of preference scores across attractiveness, charisma, and gender. Is this table easy to interpret? What might be an alternative method for obtaining a descriptive overview of the data?</li>
</ol></li>
<li><ol start="3" type="1">
<li>What is the type of test ? Which test will you use ? Explain your answer ?</li>
</ol></li>
<li><ol start="4" type="1">
<li>State the null and alternate huypothesis.</li>
</ol></li>
<li><ol start="5" type="1">
<li>Conduct the test . Report will you accept the Null hypothesis?</li>
</ol></li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>